lib/policy_machine_storage_adapters/active_record/postgresql.rb

```diff
@@ -2,6 +2,127 @@
 
 module PolicyMachineStorageAdapter
   class ActiveRecord
+    class PolicyElementAssociation
+      def self.all_accessible_objects(associations, root_id:, filters: {})
+        pea_ids = associations.pluck(:id)
+        return [] if pea_ids.empty?
+
+        query = <<-SQL
+          id IN (
+            WITH RECURSIVE ancestor_scope AS (
+            SELECT
+                id AS child_id,
+                id AS parent_id
+            FROM policy_elements
+            WHERE id = ?
+
+            UNION ALL
+
+            SELECT
+                assignments.child_id,
+                assignments.parent_id
+            FROM assignments
+            JOIN ancestor_scope
+                ON assignments.child_id = ancestor_scope.parent_id
+            ),
+
+            leaf_ancestors AS (
+            SELECT DISTINCT
```

Comment:
Do we need these distincts? In particular, I'm curious if postgres handles this duplication for us while recursing and leaves us with a performance boost if we remove the distinct because we wouldn't have to do this (potentially) redundant unique on every iteration of the recursion. Same comment below.

Reply:
These distincts are intentional for performance reasons.

The key here is the use of `UNION ALL` in the recursive CTEs. Using `UNION ALL` does _not_ enforce uniqueness, while using `UNION` does. Using `UNION` has worse performance and using `UNION ALL` leaves duplicates. By selecting distinct records _outside_ the query, we get the performance benefits of not uniqueness checking on each recursion while still getting a unique set. Having a unique set is useful because we won't have to duplicate work in recursive queries using these results.